### ISSUE TITLE
Updating containment and snap of existing panel?

### ISSUE THREAD

**Issue by fq-selbach** [Flyer53/jsPanel4#82]
Is it possible to update the snap and containment settings for an existing panel?
I've tried: `panel.options.dragit.containment = 0` to activate it but behavior of the panel does not change (still overflows). Same for 'snap' options.

The reason I'm asking is because I have 2 containers. One where overflow is allowed and snapping disabled and one where containment is 0 and snapping is enabled. Now I want to move a panel from one container to the other but the snap and containment options are bound to the panel and seem to be fixed after the panel was created.

**Comment by Flyer53** [comment]
That's not exactly built-in.
But the following procedure might work. The **dragit** interaction is setup by a mehtod of the global object `jsPanel`. You could call this method again with new parameters. Below is an example to copy/paste to your dev tools
```javascript
// create panel and store reference to it
var p = jsPanel.create({
    dragit: {
        containment: 0
    }
});

setTimeout(function () {
    // timeout for demo only

    // new dragit configuration
    // start, drag and stop callbacks must be included as either empty array or array of function(s)
    var newDragitConfig = {
        containment: 50,
        start: [],
        drag: [
            function (panel, position, event) {
                console.log(event.type);
            }
        ],
        stop: [],
        snap: {
            sensitivity: 70,
            trigger: 'panel'
        }
    };

    // overwrite dragit setting of panel's options property
    p.options.dragit = newDragitConfig;

    // reset dragit interaction
    jsPanel.dragit(p, newDragitConfig);

}, 10000);
```
**Notes:**
+ The `start`, `drag` and `stop` callbacks **must** be included in the new config. Either as empty array or as array of functions. Otherwise you get an error.
+ A `snap` config with at least the default settings shown above must be in the new config.
+ It seems to work, although there is a `TypeError` when you drag away a snapped panel.
+ Depending on your used options other issues might come up.

If it's not to complicated ... give it a try 😏

**Comment by Flyer53** [comment]
Btw: I uploaded an updated alpha version: https://alpha.jspanel.de/downloads/jspanel-4.6.0-alpha.zip

**Comment by fq-selbach** [comment]
Great I'll try that :-) Whats in the new update? ^^

**Comment by Flyer53** [comment]
Basically a fix in the tooltip extension.
And a few minor issues fixed ...

**Comment by fq-selbach** [comment]
So I tried the following on a panel that had no containment and snap enabled:
```
p.options.dragit.containment = 0;
p.options.dragit.snap = false;
jsPanel.dragit(p, p.options.dragit);
```
After this the containment works 👍 but the panel still snaps :-(

**Comment by Flyer53** [comment]
In this case `p.options.dragit.snap = false;` does not work for whatever reason.
Try
```javascript
var p = jsPanel.create({
    dragit: {
        snap: true
    }
});

setTimeout(function () {
    var newDragitConfig = {
        containment: 0,
        start: [],
        drag: [function(panel, pos, e){
            console.log(e.type);
        }],
        stop: []
    };

    p.options.dragit = newDragitConfig;

    jsPanel.dragit(p, newDragitConfig);

}, 10000);
```
Note that `snap` is simply omitted and not set to `false`. Seems to make adifference...

I guess you won't a solution that works without glitches.
Reconfiguring options of existing panels is on my list. But you're the first one asking for it and so far I didn't dare to look for all the necessary changes ;)

**Comment by fq-selbach** [comment]
I've tried `delete p.options.dragit.snap` as well but it did not work either :-(

**Comment by Flyer53** [comment]
Did you try the exact copy of my example code?

**Comment by fq-selbach** [comment]
Now I did, same result.
After the update `p.options.dragit.snap` shows `undefined` in both cases but snap is still working.

**Comment by Flyer53** [comment]
Well ... then I'm out of ideas at the moment.
Yep, I tried only with the alpha as well.
I guess you'll have to hope for a future update concerning this feature request.

**Comment by fq-selbach** [comment]
I did some experimenting and replaced the following lines:
```
//if (opts.snap) {
if (e.target.parentNode.parentNode.parentNode.options && e.target.parentNode.parentNode.parentNode.options.dragit.snap) {
```
At:
https://github.com/Flyer53/jsPanel4/blob/master/dist/jspanel.js#L947

That seems to work ^^

**Comment by Flyer53** [comment]
As said, changing options of existing panels is on my ToDo list. But I won't start this until the v4.6.0 final is out. I guess implementing this feature will require substantial changes ... so I can't promise anything ...

**Comment by fq-selbach** [comment]
I totally understand, anyway support has been great so far :-) 👍 .

In the meantime I'm getting closer to understanding whats going on ^^.
After the 2nd call of `jsPanel.dragit(p, newDragitConfig)` the `dragElmt` method is called twice on each drag, first with the old options then with the new. If I can find a way to shutdown the old one ... :-D

**Comment by Flyer53** [comment]
A new approach:
```javascript
var panel = jsPanel.create({
    dragit: {
        containment: 30,
        snap: true
    }
});

setTimeout(function() {

    var handles = jsPanel.defaults.dragit.handles.split(',');

    handles.forEach(function (handle, index, array) {

        handle = panel.querySelector(handle.trim());

        // clone will not have event listeners bound via addEventListener()
        var clone = handle.cloneNode(true);

        var parent = handle.parentElement;

        handle.parentElement.removeChild(handle);
        parent.prepend(clone);

    });

    jsPanel.dragit(panel, {containment:0, start:[], drag:[], stop:[]});

},10000);
```

**Comment by fq-selbach** [comment]
If I understand right this will remove the old drag handlers and add new ones? Sounds like a plan ^^

**Comment by Flyer53** [comment]
To be exact it clones the drag handles (event handlers are not copied in this process), removes the handles and then adds the clones again.

I think I can refine this some more...

**Comment by Flyer53** [comment]
Next stage 😏 

Let's use `jsPanel.extend()` to extend a panel instance with the new method
`reconfigureDragit(config, callback).`

I refined the method a bit and the order in which logo and title reappear is ok now.
The object for a new dragit config can be just as in a normal panel config.
**Limitations:**
+ It works only with the default drag handles (logo, title, footer).
+ It cannot handle issues like wrong colors in a minimized panel or similar things. That needs to be dealt with in an `onminimized` callback for example

**Note:** There seems to be a bug with images as headerLogo (unrelated to this topic).

Ok, here comes the code ... and good luck 😏 
```javascript
// extend panel with method to reconfigure dragit
jsPanel.extend({
    reconfigureDragit: function(config, callback) {

        ['start', 'drag', 'stop'].forEach(function (item) {
            if (config[item]) {
                if (typeof config[item] === 'function') {
                    config[item] = [config[item]];
                }
            } else {
                config[item] = [];
            }
        });
        if (config.snap) {
            if (typeof config.snap === 'object') {
                config.snap = Object.assign({}, jsPanel.defaultSnapConfig, config.snap);
            } else {
                config.snap = jsPanel.defaultSnapConfig;
            }
        }

        var logo = this.header.querySelector('.jsPanel-headerlogo'),
            title = this.header.querySelector('.jsPanel-titlebar'),
            ftr = this.querySelector('.jsPanel-ftr');

        var logoClone = logo.cloneNode(true),
            titleClone = title.cloneNode(true),
            ftrClone = ftr.cloneNode(true);

        [logo, title, ftr].forEach(function (item) {
            if (item) {
                item.parentElement.removeChild(item);
            }
        });

        this.headerbar.prepend(titleClone);
        this.headerbar.prepend(logoClone);
        this.insertBefore(ftrClone, this.querySelector('.jsPanel-resizeit-handle'));

        jsPanel.dragit(this, config);

        if (callback) {
            callback.call(this, this);
        }

        return this;

    }
});

// USAGE:
// create a test panel
var p = jsPanel.create({
    dragit: {
        containment: 30,
        snap: true
    }
});

// call panel.reconfigureDragit()
setTimeout(function(){
    p.reconfigureDragit({
        containment: 0
    }, function(panel){
        console.log('do something with panel ...');
    });
}, 10000);
```

**Comment by fq-selbach** [comment]
Thanks a lot! Unfortunately I'm out of time for this week, but will have a look on Monday again :-) 👍

**Comment by fq-selbach** [comment]
So I've been playing around a bit with the 'reconfigureDragit' method. It is working up to a certain point, but it seems to break several things. We mentioned the 'theme' settings already, but for example the buffered pre-snap size gets deleted as well and there might be more things I did not even notice :-(
Btw it also crashes when the panel has no header.

I think the better solution would be to track the event listeners for `dragElmt` properly and simply remove them before adding new ones.
Right now I'm using a pretty dirty hack to suppress the 'outdated' listeners. In the dragit method I keep track of updates:
```
// keep track of drag handler updates
if (!elmt.dragHandlerVersion) elmt.dragHandlerVersion = 1;
else elmt.dragHandlerVersion++;
var thisVersion = elmt.dragHandlerVersion;
```
and later in the `dragElmt` method I check if it's outdated:
```
if (thisVersion != elmt.dragHandlerVersion){
    return;
}
```
It would be better to simply use removeEventListener somehow, but so far my attempts all failed.

**Comment by Flyer53** [comment]
@fq-selbach 
Sorry that I didn't answer to this so long ...
To be able to update (almost) **all** panel options of existing panels is something circling my mind for a while now. But that would require substantial code rewrites I probably won't do in v4.x. If I start a version 5 it will be part of it. However, for the time being I can't give you an estimate when that will be ...

**Comment by Flyer53** [comment]
@fq-selbach 
Hold on ...
I'm experimenting with some code changes allowing to update `dragit.containment`/`dragit.snap` configs of existing panels.
My tests are working already for `dragit.containment`. I hope I get it done for `dragit.snap` as well ....

I'll keep you posted ...

**Comment by Flyer53** [comment]
@fq-selbach 
I just uploaded v4.9.0-alpha.3 to https://alpha.jspanel.de/
**Almost all dragit parameters are editable for existing panels.**
Docs are [here](https://alpha.jspanel.de/api.html#options/dragit)
[Download](https://alpha.jspanel.de/downloads/jspanel-4.9.0-alpha.3.zip)

And I will of course continue with similar changes for other jsPanel options ...

**Comment by Flyer53** [comment]
@fq-selbach 
Already updated to v4.9.0-alpha.4 where also almost all **resizeit parameters** and option **maximizedMargin** are editable for existing panels.

[Download v4.9.0-alpha.4](https://alpha.jspanel.de/downloads/jspanel-4.9.0-alpha.4.zip)

I think it now contains everything you asked for in your original posting ...

**Comment by fq-selbach** [comment]
Hi Stefan,

thanks for the info and updates. jsPanel is running pretty stable for us right now using the "workarounds"/tweaks I've mentioned above (and in some other issues). Unfortunately this also means I cannot easily upgrade to the newest version because I need to transfer the other modifications as well and it seems that many parts of the code have changed since 4.6.0 so this will take some time.
I've made a snapshot of the original 4.6.0 and our modified 4.6.0 to track those changes and will try to fix this issue at some (not too distant) point ^^.

Regards,
Florian

**Comment by Flyer53** [comment]
@fq-selbach 
Hi Florian,
Absolutely understandable! And it's probably a good idea for you wait some more since I'm in the process of adding a `drop` feature to the dragit interaction. I think you once mentioned using a drop feature ... so just be patient a bit more ...

Regards,
Stefan

**Comment by Flyer53** [comment]
Updated to https://alpha.jspanel.de/downloads/jspanel-4.9.0-alpha.8.zip

**Comment by Flyer53** [comment]
Implemented in v4.9.0 release
